src/vsock/src/vhost_user_vsock_thread.rs

```diff
@@ -0,0 +1,549 @@
+                let mut new_vsock_conn = VsockConnection::new_local_init(
+                    unix_stream,
+                    VSOCK_HOST_CID,
+                    local_port,
+                    self.guest_cid,
+                    peer_port,
+                    self.get_epoll_fd(),
+                );
+                new_vsock_conn.rx_queue.enqueue(RxOps::Request);
+                new_vsock_conn.set_peer_port(peer_port);
+
+                // Add connection object into the backend's maps
+                self.thread_backend
+                    .conn_map
+                    .insert(conn_map_key, new_vsock_conn);
+
+                self.thread_backend
+                    .backend_rxq
+                    .push_back(ConnMapKey::new(local_port, peer_port));
+
+                // Re-register the fd to listen for EPOLLIN and EPOLLOUT events
+                Self::epoll_modify(
+                    self.get_epoll_fd(),
+                    fd,
+                    epoll::Events::EPOLLIN | epoll::Events::EPOLLOUT,
+                )
+                .unwrap();
+            } else {
+                // Previously connected connection
+                let key = self.thread_backend.listener_map.get(&fd).unwrap();
+                let vsock_conn = self.thread_backend.conn_map.get_mut(&key).unwrap();
+
+                if evset == epoll::Events::EPOLLOUT {
+                    // Flush any remaining data from the tx buffer
+                    match vsock_conn.tx_buf.flush_to(&mut vsock_conn.stream) {
+                        Ok(cnt) => {
+                            vsock_conn.fwd_cnt += Wrapping(cnt as u32);
+                            vsock_conn.rx_queue.enqueue(RxOps::CreditUpdate);
+                            self.thread_backend.backend_rxq.push_back(ConnMapKey::new(
+                                vsock_conn.local_port,
+                                vsock_conn.peer_port,
+                            ));
+                        }
+                        Err(e) => {
+                            dbg!("Error: {:?}", e);
+                        }
+                    }
+                    return;
+                }
+
+                // Unregister stream from the epoll, register when connection is
+                // established with the guest
+                Self::epoll_unregister(self.epoll_file.as_raw_fd(), fd).unwrap();
+
+                // Enqueue a read request
+                vsock_conn.rx_queue.enqueue(RxOps::Rw);
+                self.thread_backend
+                    .backend_rxq
+                    .push_back(ConnMapKey::new(vsock_conn.local_port, vsock_conn.peer_port));
+            }
+        }
+    }
+
+    /// Allocate a new local port number.
+    fn allocate_local_port(&mut self) -> Result<u32> {
+        // TODO: Improve space efficiency of this operation
+        // TODO: Reuse the conn_map HashMap
+        let mut alloc_local_port = self.local_port.0;
```

Comment:
Maybe we can save the last port used and restart from it, instead of restarting from 0 every time, but I saw your TODO item, so we can do later.

Reply:
Another TODO item can be a test to check if we use all the ports we receive the expected error.

Reply:
> Maybe we can save the last port used and restart from it, instead of restarting from 0 every time, but I saw your TODO item, so we can do later.

It actually starts from the last allocated port, although this may change slightly if I can get the space efficient solution working.

Reply:
> Another TODO item can be a test to check if we use all the ports we receive the expected error.
Ok. Not sure if I should do this right away? The metadata to track the used ports would be atleast 2^32 times the size of the hash structure in the hashset.

Reply:
> > Maybe we can save the last port used and restart from it, instead of restarting from 0 every time, but I saw your TODO item, so we can do later.
> 
> It actually starts from the last allocated port, although this may change slightly if I can get the space efficient solution working.

Right :-) I was a bit confused...

Reply:
> > Another TODO item can be a test to check if we use all the ports we receive the expected error.
> 
> Ok. Not sure if I should do this right away? The metadata to track the used ports would be atleast 2^32 times the size of the hash structure in the hashset.

Yep, so let's do it later.